### DISCUSSION TITLE
sqlalchemy 2.0 - query property typing

### DISCUSSION THREAD

**Discussion by xmatthias** [sqlalchemy/sqlalchemy#9338] [Usage Questions] [answered]
using Sqlalchemy 2.0.x ORM - in combination with mapped columns and mypy 1.0 - what's the correct way to type the "query" property manually (assignment of the actual property happens later as part of initialization).

``` python
from typing import ClassVar, Any
from sqlalchemy import create_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, scoped_session, sessionmaker


class Base(DeclarativeBase):
    pass


class Model1(Base):
    query: ClassVar[Any]   # How to type this "manually" ? 
    id: Mapped[int]
    name: Mapped[str]


def db_setup(db_url: str, **kwargs):

    ...
    engine = create_engine(db_url, future=True, **kwargs)
    session = scoped_session(sessionmaker(bind=engine, autoflush=False))

    Model1.query = session.query_property()


reveal_type(Model1.query)

```

Based on inspection through vscode - the correct type seems to be `ClassVar[_QueryDescriptorType]` - which makes subsequent usages of `Model1.query.filter().all()` return the correct type.

I'm currently importing as follows: `from sqlalchemy.orm.scoping import _QueryDescriptorType`

However, i suspect this type is not meant to be used this way (and therefore not safe - as it may change at any time) - as it's not exported and marked as "private" by convention (starting with `_`).

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think you can type it with `Query` directly. Something like `query: ClassVar[Query['Model1']]` should work

**Reply by CaselIT** [reply]
Even better, this works

```py
class Base(DeclarativeBase):
    query: ClassVar[Query[Self]]
```

**Reply by xmatthias** [reply]
If you can stricktly stick to python 11 - sure.
for any project trying to support anything below - unfortunately using `Self` is not an option. (that's not a problem though - using string-typed class-name would be fine).

unfortunately, `query: ClassVar[Query['Model1']]` is also not really working without introducing another problem


``` python
from typing import ClassVar

from sqlalchemy import create_engine, desc
from sqlalchemy.orm import DeclarativeBase, Mapped, Query, scoped_session, sessionmaker


class Base(DeclarativeBase):
    pass


class Model1(Base):
    query: ClassVar[Query['Model1']]
    id: Mapped[int]
    name: Mapped[str]


def db_setup(db_url: str, **kwargs):

    ...
    engine = create_engine(db_url, future=True, **kwargs)
    session = scoped_session(sessionmaker(bind=engine, autoflush=False))

    Model1.query = session.query_property()


reveal_type(Model1.name)
reveal_type(Model1.query)
aaa = Model1.query.order_by(desc(Model1.id)).all()
reveal_type(aaa)

```

running mypy over this results in the following:

```
sqlalchemy_test.py:23: error: Incompatible types in assignment (expression has type "_QueryDescriptorType", variable has type "Query[Model1]")  [assignment]
sqlalchemy_test.py:26: note: Revealed type is "sqlalchemy.orm.attributes.InstrumentedAttribute[builtins.str]"
sqlalchemy_test.py:27: note: Revealed type is "sqlalchemy.orm.query.Query[sqlalchemy_test.Model1]"
sqlalchemy_test.py:29: note: Revealed type is "builtins.list[sqlalchemy_test.Model1]"
```

Apparently, `_QueryDescriptorType` (what's returned by `query_property()`) and Query are incompatible on assignment, so this is apparently not the correct solution (though the query result itself is properly typed).

**Reply by CaselIT** [reply]
> for any project trying to support anything below - unfortunately using Self is not an option. (that's not a problem though - using string-typed class-name would be fine).

you can use it with typing-extension.

> Apparently, _QueryDescriptorType (what's returned by query_property()) and Query are incompatible on assignment, so this is apparently not the correct solution (though the query result itself is properly typed).

It appear so, I think the best option here is to change the return type of `query_property` to make it compatible with `Query`.
@zzzeek what do you think?

**Reply by zzzeek** [reply]
if `ClassVar[Query[Self]]` why arent we using that?   i dont know what the proposal is otherwise.

**Reply by xmatthias** [reply]
well `session.query_property()` is not returning that, but `_QueryDescriptorType` - so the above code is violating mypy rules.

**Reply by CaselIT** [reply]
My proposal was to either "lie" to the type checker so that `session.query_property()` is typed to return `Query[Any]` or to make the descript type public.

**Reply by zzzeek** [reply]
ok im not doing too well with following specifics today, if we can put a file in mypy/plain_files/ that shows all the cases and affirms the types we want to see, let's start with that and then whatever works in scoping.py we do.  hows that?

**Reply by CaselIT** [reply]
I've created https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4460

I've gone with exporting the descriptor, since it seems it already did figure out the correct type to attach to the query

**Reply by xmatthias** [reply]
@CaselIT i'm not too familiar with how the sqlalchemy contribution process works - is the above commit available somehow as PR / branch to check out and test to see if it works for my case?

**Reply by CaselIT** [reply]
If you follow the gerrit link you can find a download patch button that you can use to apply it locally

**Reply by xmatthias** [reply]
i guess best let me know once it's available in a user/developer friendly way - either as branch in this repo, or merged / released.

**Reply by CaselIT** [reply]
you can download a patch following this link https://gerrit.sqlalchemy.org/changes/sqlalchemy%2Fsqlalchemy~4460/revisions/2/patch?zip

you can then unzip and apply with git apply

**Reply by xmatthias** [reply]
Great job 👍 
works fine with 2.0.5

**Reply by CaselIT** [reply]
great, thanks for confirming it now works
